### PR TITLE
Make disk_cache persist across runs

### PR DESCRIPTION
**Pull request by lakesare**

### CODE REVIEW

File: sciunit/models/backends.py [hunk 1]
```diff
@@ -69,17 +69,22 @@ def init_memory_cache(self) -> None:
...
-            else:
-                self.disk_cache_location.unlink()
-        except Exception:
-            pass
-        self.disk_cache_location = Path(tempfile.mkdtemp()) / "cache"
+        from_kwargs = kwargs.get("location")
+        if isinstance(from_kwargs, Path) or isinstance(from_kwargs, str):
+            location = str(from_kwargs)
+        else:
+            location = Path(tempfile.mkdtemp()) / "cache"
```
Comment (lakesare) [inline sciunit/models/backends.py:78]: @rgerkin, as mentioned in the comment, we'll need to add `config.set_to_disk()` for the commented-out version to work.

Before I commence with that, however, how about another idea.
We can use `Path(os.getcwd()) / 'cache'`, which will create the `cache` file in the folder the user executes their script from.  
This way we don't need to mess with the config, and the user keeps direct control over the size of their cache file.  
Do you see some drawbacks to this approach?

**Review by rgerkin — CHANGES_REQUESTED** [review]

File: sciunit/models/backends.py [hunk 2]
```diff
@@ -69,17 +69,22 @@ def init_memory_cache(self) -> None:
...
-            # Cleanup old disk cache files
-            if self.disk_cache_location.is_dir():
-                self.disk_cache_location.rmdir()
-            else:
-                self.disk_cache_location.unlink()
-        except Exception:
-            pass
-        self.disk_cache_location = Path(tempfile.mkdtemp()) / "cache"
+        from_kwargs = kwargs.get("location")
+        if isinstance(from_kwargs, Path) or isinstance(from_kwargs, str):
```
Comment (rgerkin) [inline sciunit/models/backends.py:75]: @lakesare the second argument of instance can be a tuple of types, fyi

File: sciunit/models/backends.py (hunk 1 above)
Comment (rgerkin) [reply sciunit/models/backends.py:78]: Yes I prefer that, and it will persist until the user deletes.  It please do that.

**Review by rgerkin — CHANGES_REQUESTED** [review]
See slack and below.

File: sciunit/models/backends.py (hunk 1 above)
Comment (lakesare) [reply sciunit/models/backends.py:78]: To set the record straight - we don't want `tempfile` to handle the cache location anymore (and thus we don't need to store any paths in `$HOME/.sciunit/config.json`), we want a persistent cache in `$HOME/.sciunit/cache`.
> Yes, let’s do the .sciunit location (something like $HOME/.sciunit/cache) for disk_cache=True

Comment (rgerkin) [reply sciunit/models/backends.py:78]: I suppose we could still allow it with some special path location argument (e.g. "_temp_") but I'm not sure what the use case would be for this, so I think you can get rid of it.

File: sciunit/unit_test/backend_tests.py [hunk 3]
```diff
@@ -27,10 +27,34 @@ def test_backends_init_caches(self):
...
+        backend = Backend()
+        backend.model = myModel
+        backend.init_backend(use_disk_cache=True, use_memory_cache=False)
+        self.assertTrue(backend.disk_cache_location.endswith(".sciunit/cache"))
+
+        # Manually set disk_cache location (a string)
+        myModel = Model()
+        backend = Backend()
+        backend.model = myModel
+        backend.init_backend(use_disk_cache="/some/good/path", use_memory_cache=False)
```
Comment (rgerkin) [inline sciunit/unit_test/backend_tests.py:42]: Can this fake path possibly work?  Can you instead use the `tempfile` module to just explicitly create an real temporary path?  I assume that `init_backend` doesn't actually try to write to this location (so it doesn't matter if it is real or not), but in that case it isn't much of a test of whether a manual location can work.  You should instead have each of these cases actually try writing data to that cache, checking if it is there, etc.

File: sciunit/utils.py [hunk 4]
```diff
@@ -468,6 +468,22 @@ def write(self, s: str) -> None:
         if s.startswith("[") and s.endswith("]"):
             super(MockDevice, self).write(s)
 
+class TmpTestFolder():
+    '''A class for creating and deleting a folder in "./unit_test/".
+    '''
+
+    path = Path(__file__).parent / "unit_test" / "delete_after_tests"
```
Comment (rgerkin) [inline sciunit/utils.py:475]: I think we are moving to doing everything inside HOME/.sciunit, so maybe the temp test folder should be there as well, i.e. HOME/.sciunit/unit_test.  A user could theoretically be running code from a folder in which they don't have write access, for example.

**Review by rgerkin — CHANGES_REQUESTED** [review]
A few changes suggested

File: sciunit/utils.py (hunk 4 above)
Comment (lakesare) [reply sciunit/utils.py:475]: > A user could theoretically be running code from a folder in which they don't have write access, for example.

And by the user you mean the sciunit developer who runs the tests? (we only use this util for tests)

Would you prefer your temporary test files to be in `HOME/.sciunit`, somewhere outside of your git tracker?
If something bad happens and test files don't get deleted I'd certainly like to see them as my unstaged files.
If something doesn't work I'd want to uncomment the line that deletes the file, and check the file myself (without going to my HOME folder).
Do these concerns sound unrelatable?

Comment (rgerkin) [reply sciunit/utils.py:475]: I think there are probably arguments both ways so I think it is fine as you have it.

**Pull request merged**